rclcpp/include/rclcpp/allocator/allocator_common.hpp

```diff
@@ -39,61 +39,59 @@ void * retyped_allocate(size_t size, void * untyped_allocator)
   return std::allocator_traits<Alloc>::allocate(*typed_allocator, size);
 }
 
-template<typename T, typename Alloc>
+template<typename Alloc>
 void retyped_deallocate(void * untyped_pointer, void * untyped_allocator)
 {
   auto typed_allocator = static_cast<Alloc *>(untyped_allocator);
   if (!typed_allocator) {
     throw std::runtime_error("Received incorrect allocator type");
   }
-  auto typed_ptr = static_cast<T *>(untyped_pointer);
+  auto typed_ptr = static_cast<typename Alloc::value_type *>(untyped_pointer);
   std::allocator_traits<Alloc>::deallocate(*typed_allocator, typed_ptr, 1);
 }
 
-template<typename T, typename Alloc>
+template<typename Alloc>
 void * retyped_reallocate(void * untyped_pointer, size_t size, void * untyped_allocator)
 {
   auto typed_allocator = static_cast<Alloc *>(untyped_allocator);
   if (!typed_allocator) {
     throw std::runtime_error("Received incorrect allocator type");
   }
-  auto typed_ptr = static_cast<T *>(untyped_pointer);
+  auto typed_ptr = static_cast<typename Alloc::value_type *>(untyped_pointer);
   std::allocator_traits<Alloc>::deallocate(*typed_allocator, typed_ptr, 1);
   return std::allocator_traits<Alloc>::allocate(*typed_allocator, size);
 }
 
+}  // namespace allocator
 
-// Convert a std::allocator_traits-formatted Allocator into an rcl allocator
-template<
-  typename T,
-  typename Alloc,
-  typename std::enable_if<!std::is_same<Alloc, std::allocator<void>>::value>::type * = nullptr>
+// Deprecated: Generic converter from C++ allocator into RCL allocator.
+// This allocator overallocates memory by 100x and invokes UB on free,
+// see #1254.
+template<typename Alloc>
```

Comment:
Honestly, if this thing is so broken, I think we should just remove it.  No reason to keep time-bombs around for others to find.

Reply:
+1, I totally want to remove it. tlsf_cpp and demo_nodes still use this overload, though. My plan is to submit this PR, after which I can add overloads for get_rcl_allocator to tlsf_cpp and demo_nodes, and then drop this overload.

Reply:
Yeah I think we should remove it. There's other allocator work that needs to be done, and in general we will probably end up dropping all the allocator stuff in favor the polymorphic allocators from C++17 when we can.

Reply:
Seems we agree. Shall we submit this change and I'll send a follow-up PR to be submitted when dependencies are resolved?

Reply:
Generally my preference is to do it all at once (remove this API and update the packages using it), but you can do it in any order that works for you. We just need all tests passing and no compiler warnings for each set of pull requests merged.